src/Plutarch/Test/QuickCheck/Instances.hs

```diff
@@ -439,6 +456,243 @@ instance Function Address where
+ is somewhat limited, but otherwise would require UTF-8 encoding as part of
+ the generator. It would also significantly complicate shrinks: we would have
+ to re-encode, shrink, then decode again.
+
+ @since 2.1.3
+-}
+instance Arbitrary TokenName where
+  {-# INLINEABLE arbitrary #-}
+  arbitrary =
+    TokenName <$> do
+      ASCIIString name <- arbitrary
+      pure
+        . toBuiltin @ByteString
+        . fromList
+        . fmap (fromIntegral . ord)
+        $ name
+  {-# INLINEABLE shrink #-}
+  shrink (TokenName name) = do
+    let name' =
+          ASCIIString
+            . fmap (chr . fromIntegral)
+            . toList
+            . fromBuiltin
+            $ name
+    ASCIIString name'' <- shrink name'
+    pure
+      . TokenName
+      . toBuiltin @ByteString
+      . fromList
+      . fmap (fromIntegral . ord)
+      $ name''
+
+-- | @since 2.1.3
+deriving via BuiltinByteString instance CoArbitrary TokenName
+
+-- | @since 2.3.1
+instance Function TokenName where
+  {-# INLINEABLE function #-}
+  function = functionMap (coerce @_ @BuiltinByteString) coerce
+
+{- | This generates a phase-1 valid 'Value' meant to represent an /amount/,
+ rather than a /delta/. Thus, all of the following will hold for all generated
+ values and shrinks:
+
+ - Entries whose 'CurrencySymbol' and 'TokenName' match are unique.
+ - \'Outer\' map entries are sorted by 'CurrencySymbol'.
+ - \'Inner\' map entries are sorted by 'TokenName'.
+ - An ADA entry always exists.
+ - All entries have positive amounts.
+
+ While it is possible to have a normalized 'Value' with a zero ADA entry, no
+ such entries are currently generated.
+
+ 'Value's are shrunk only structurally: this means that we can \'drop\'
+ entries, but not shrink existing ones in either their \'inner map\' keys or
+ \'inner map\' values.
+
+ @since 2.1.3
+-}
+instance Arbitrary Value where
+  {-# INLINEABLE arbitrary #-}
+  arbitrary =
+    Value . AssocMap.fromList <$> do
+      (Sorted cs) <- arbitrary
+      case cs of
+        [] -> do
+          adaEntry <- mkEntry ""
+          pure [adaEntry]
+        (sym : _) ->
+          if sym == ""
+            then traverse mkEntry cs
+            else do
+              adaEntry <- mkEntry ""
+              entries <- traverse mkEntry cs
+              pure $ adaEntry : entries
+    where
+      mkEntry ::
+        CurrencySymbol ->
+        Gen (CurrencySymbol, AssocMap.Map TokenName Integer)
+      mkEntry sym = do
+        (SortedNonEmpty tns) <- arbitrary
+        kvs :: [(TokenName, Positive Integer)] <-
+          traverse (\tn -> (tn,) <$> arbitrary) tns
+        pure (sym, AssocMap.fromList . coerce $ kvs)
+  {-# INLINEABLE shrink #-}
+  shrink (Value inner) =
+    Value <$> do
+      let inner' = fmap PlutusTx.toList <$> PlutusTx.toList inner
+      case inner' of
+        [] -> error "Shrinker for Value: impossible situation: given empty Value to shrink"
```

Comment:
shouldn't empty value just return empty value? Is it because empty value is not onchain-valid?

Reply:
The whole idea is that we always keep the ADA entry regardless of shrinks.